napari/components/viewer_model.py

```diff
@@ -1000,6 +1000,69 @@ def add_vectors(
         self.add_layer(layer)
         return layer
 
+    def _add_layer_data(
+        self, data, meta: dict = None, layer_type: str = 'image'
+    ):
+        """Add arbitrary layer data to the viewer.
+
+        Primarily intended for usage by reader plugin hooks.
+
+        Parameters
+        ----------
+        data : Any
+            Data in a format that is valid for the corresponding `add_*` method
+            of the specified ``layer_type``.
+        meta : dict, optional
+            Dict of keyword arguments that will be passed to the corresponding
+            `add_*` method.  MUST NOT contain any keyword arguments that are
+            not valid for the corresponding method.
+        layer_type : str
+            Type of layer to add.  MUST have a corresponding add_* method on
+            on the viewer instance.
+
+        Raises
+        ------
+        ValueError
+            If ``layer_type`` is not one of the recognized layer types.
+        TypeError
+            If any keyword arguments in ``meta`` are unexpected for the
+            corresponding `add_*` method for this layer_type.
+
+        Examples
+        --------
+        A typical use case might be to upack a tuple of layer data with a
+        specified layer_type.
+
+        >>> viewer = napari.Viewer()
+        >>> data = (
+        ...     np.random.random((10, 2)) * 20,
+        ...     {'face_color': 'blue'},
+        ...     'points',
+        ... )
+        >>> viewer._add_layer_data(*data)
+
+        """
+
+        try:
+            add_method = getattr(self, 'add_' + layer_type)
+        except AttributeError:
+            valid_types = {
```

Comment:
ahh.  good catch.  If possible, I'd prefer to *not* have to specify every valid layer type, just because it's one more thing to have to remember if we add new `Layer` types later... but then again we may eventually add new public `add_*` methods to the viewer as well, so simply excluding `layer_type == " layer"` could also break in the future.  Do we have a "single source of truth" somewhere in the codebase for valid layer names?

Reply:
how about something like this (in addition to making sure an `add_*` method is available)
```python
valid_layer_types = [subclass.__name__.lower() for subclass in Layer.__subclasses__()]
```

Reply:
I agree that we should avoid adding something else that needs to be manually curated when adding/modifying the layer classes. This would definitely not be the intended usage, but there is a dictionary containing valid layers in the [vispy utils](https://github.com/napari/napari/blob/b32501198c52e3145dca1fb2feeac284227013a7/napari/_vispy/utils.py#L9) (used to map layer models to views). Maybe somebody else knows a better way!

Reply:
> how about something like this (in addition to making sure an add_* method is available)

Sorry, I missed this! That looks good to me!

Reply:
realized I actually need to make it recursive in order to catch things like `Labels` (which is subclassed from `Image`)...  so I made a convenience method in utils, and added a new module-level `layers.names` variable that can serve as a source-of-truth if everyone is ok with that.

Reply:
I love the recursive subclasses approach! Probably useful in other parts of the codebase!